README.md
# SwitchFlit

A plug-and-play quick-switch UI for SilverStripe DataObjects.

![screenshot](https://cloud.githubusercontent.com/assets/242621/19829330/e831df5a-9e39-11e6-9e6c-413f7b6d7ed4.png)

## Premise

Quick-Switcher UIs are super useful for power users, and available today in a variety of applications and forms, such as Slack, Sublime Text, and Spotlight. With SwitchFlit, you can specify a DataObject's Name and Link fields and immediately enable a basic Quick-Switcher in your web app.

## Installation / Configuration

`composer require cheddam/silverstripe-switchflit`

See the [installation guide](docs/en/installation.md) for basic setup.

## Technologies Used

SwitchFlit uses [Vue](https://vuejs.org) to power its UI, assisted by [Fuse.js](http://fusejs.io) for fuzzy-search.

## Contributing

See `CONTRIBUTING.md`.


CHANGELOG.md
# Changelog

## [1.0.0](https://github.com/Cheddam/SwitchFlit/tree/v1.0.0)

The first stable release!

- Added DataObject Alias parameter for displaying an alternative DataObject name in the UI
- Changed `result-link` to `result-details` to future-proof the API slightly
- Updated the preview image in the README.

## [1.0.0 Beta 2](https://github.com/Cheddam/SwitchFlit/tree/v1.0.0-beta.2) (2016-10-14)

The second beta build! This release focusses on improving the default UI and interactions.

- Added mouse interactions
- Added basic docs about customising style and structure
- Improved visibility of loading / error states
- Improved design, should function better outside the default Simple theme

## [1.0.0 Beta 1](https://github.com/Cheddam/SwitchFlit/tree/v1.0.0-beta.1) (2016-10-01)

Welcome to the first beta build! Some major strides have been made since 0.9, and we are well on track to hitting 1.0 by the end of October.

- Added `WithCustomQuery` interface to allow custom filtering and other behaviour on records (see new docs)
- Added switching between results (use the arrow keys!)
- Added testing, linting, shrinkwrapping and editorconfig to keep things ship-shape
- Improved namespacing consistency
- Improved error handling and display across front- and back-end
- Improved global hotkey (now tentatively supports iOS devices and only becomes available when SwitchFlit is ready)
- Fixed inconsistent licensing

## [0.9.0](https://github.com/Cheddam/SwitchFlit/tree/v0.9.0) (2016-08-19)

Initial release.


tests/SwitchFlitControllerTest.php
<?php

use SwitchFlit\SwitchFlitable;
use SwitchFlit\WithCustomQuery;

class SwitchFlitControllerTest extends FunctionalTest
{
    protected $usesDatabase = true;

    protected static $fixture_file = 'SwitchFlitControllerTest.yml';

    public function testReturnsJson()
    {
        $response = $this->get('/switchflit/SwitchFlitDataObject/records');

        $this->assertEquals('application/json', $response->getHeader('Content-Type'));
    }

    public function testGetAllowedRecordsForSwitchFlitableDataObject()
    {
        $expected = (object)[
            'items' => [
                (object)[
                    'title' => 'First',
                    'link' => '/sfobject/1'
                ],
                (object)[
                    'title' => 'Third',
                    'link' => '/sfobject/3'
                ],
            ],
        ];

        $response = $this->get('/switchflit/SwitchFlitDataObject/records');
        $this->assertEquals(200, $response->getStatusCode());

        $body = json_decode($response->getBody());
        $this->assertEquals($expected->items, $body->items);
    }

    public function testDoNotGetRecordsForNonSwitchFlitableDataObject()
    {
        $expected = (object)[
            'errors' => ['The class Member is not SwitchFlitable.']
        ];

        $response = $this->get('/switchflit/Member/records');

        $this->assertEquals(400, $response->getStatusCode());
        $this->assertEquals($expected, json_decode($response->getBody()));
    }

    public function testDoNotGetRecordsForNonExistentDataObject()
    {
        $expected = (object)[
            'errors' => ['The class NotReal does not exist.']
        ];

        $response = $this->get('/switchflit/NotReal/records');

        $this->assertEquals(400, $response->getStatusCode());
        $this->assertEquals($expected, json_decode($response->getBody()));
    }

    public function testDoNotGetRecordsForNonDataObject()
    {
        $expected = (object)[
            'errors' => ['The class stdClass is not a DataObject.']
        ];

        $response = $this->get('/switchflit/stdClass/records');

        $this->assertEquals(400, $response->getStatusCode());
        $this->assertEquals($expected, json_decode($response->getBody()));
    }

    public function testGetCorrectRecordsWithCustomQuery()
    {
        $expected = (object)[
            'items' => [
                (object)[
                    'title' => 'Second',
                    'link' => '/sfobject/2'
                ],
            ],
        ];

        $result = $this->get('/switchflit/SwitchFlitDataObjectWithCustomQuery/records');
        $body = json_decode($result->getBody());
        $this->assertEquals($expected->items, $body->items);
    }
}

class SwitchFlitDataObject extends DataObject implements SwitchFlitable
{
    private static $db = [
        'Name' => 'Varchar(150)',
    ];

    public function SwitchFlitTitle()
    {
        return $this->Name;
    }

    public function SwitchFlitLink()
    {
        return '/sfobject/' . $this->ID;
    }

    public function canView($member = null)
    {
        return $this->ID != 2;
    }
}

class SwitchFlitDataObjectWithCustomQuery extends DataObject implements SwitchFlitable, WithCustomQuery
{
    private static $db = [
        'Name' => 'Varchar(150)',
        'Type' => "Enum('Valid,Invalid','Invalid')",
    ];

    /**
     * @return string The title to use in SwitchFlit for this DataObject
     */
    public function SwitchFlitTitle()
    {
        return $this->Name;
    }

    /**
     * @return string The link to use in SwitchFlit for this DataObject
     */
    public function SwitchFlitLink()
    {
        return '/sfobject/' . $this->ID;
    }

    /**
     * @param \DataList $data The original DataList.
     * @return \DataList The DataList with custom filters applied.
     */
    public static function SwitchFlitQuery(\DataList $data)
    {
        return $data->filter(['Type' => 'Valid']);
    }

    public function canView($member = null)
    {
        return true;
    }
}


js/switchflit.js
/* global window, navigator, console */
const Vue = require('vue');
const Fuse = require('fuse.js');
const indefiniteArticle = require('indefinite-article');
const { fetch } = require('fetch-ponyfill')();

// eslint-disable-next-line no-new
new Vue({
  el: '#switchflit',

  props: ['dataobject', 'alias'],

  data: {
    visible: false,
    currentState: 'loading', // (loading | ready | error)
    config: {},
    states: {
      loading: {
        title: 'Loading Data',
        description: 'Currently downloading data. You may see incomplete search results.',
      },
      ready: {
        title: 'Ready',
        description: 'Everything is working as expected.',
      },
      error: {
        title: 'Load Failure',
        description: 'Failed to load data. Search is unavailable.',
      },
    },
    records: [],
    fuse: null,
    query: '',
    selectedResult: 0,
  },

  computed: {
    filteredRecords() {
      if (!this.fuse) return [];

      this.fuse.set(this.records);

      return this.fuse.search(this.query);
    },
    currentStateClass() {
      return `switchflit-${this.currentState}`;
    },
    currentStateTitle() {
      return this.states[this.currentState].title;
    },
    currentStateDescription() {
      return this.states[this.currentState].description;
    },
    queryPlaceholder() {
      return `Find ${indefiniteArticle(this.alias)} ${this.alias}...`;
    },
  },

  methods: {
    show() { this.visible = true; window.setTimeout(() => { this.$el.querySelector('input').focus(); }, 0); },
    hide() { this.visible = false; this.query = ''; },

    shiftDown() {
      if (this.selectedResult < this.filteredRecords.length - 1) {
        this.selectedResult = this.selectedResult + 1;
      }
    },
    shiftUp() {
      if (this.selectedResult > 0) {
        this.selectedResult = this.selectedResult - 1;
      }
    },
    selectResult(i) {
      this.selectedResult = i;
    },

    openResult() {
      if (this.filteredRecords.length > 0) {
        window.location = this.filteredRecords[this.selectedResult].link;
      }
    },
  },

  created() {
    fetch(`/switchflit/${this.dataobject}/records`)
      .then((response) => {
        if (!response.ok) {
          response.json().then((parsedBody) => {
            // eslint-disable-next-line no-console
            console.error(`[switchflit] Heads up, something went wrong: ${parsedBody.errors}`);

            this.currentState = 'error';
          });
        } else {
          response.json().then((parsedBody) => {
            this.config = parsedBody.config;
            this.records = parsedBody.items;

            this.fuse = new Fuse(this.records, { keys: ['title'] });

            this.currentState = 'ready';
          });
        }
      })
      .catch((error) => {
        // eslint-disable-next-line no-console
        console.log(`[switchflit] Heads up, something went wrong: ${error.message}`);

        this.currentState = 'error';
      });
  },

  ready() {
    document.addEventListener('keydown', (event) => {
      if (event.keyCode === 27) {
        this.hide();
      } else if (['MacIntel', 'iPhone', 'iPad', 'iPod'].indexOf(navigator.platform) > -1) {
        if(typeof this.config.key_combo_mac !== 'undefined' && this.config.key_combo_mac.length > 0) {
          var keys = this.config.key_combo_mac;
        }
      } else {
        if(typeof this.config.key_combo !== 'undefined' && this.config.key_combo.length > 0) {
          var keys = this.config.key_combo_mac;
        }
      }

      var matchingKeys = 0;

      keys.forEach(function(key) {
        if(parseInt(key) >= 0 && event.keyCode === parseInt(key)) {
          matchingKeys++;
        } else if (
          key.trim() === 'alt' && event.altKey ||
          key.trim() === 'ctrl' && event.ctrlKey ||
          key.trim() === 'meta' && event.metaKey ||
          key.trim() === 'shift' && event.shiftKey
        ) {
          matchingKeys++;
        }
      });

      if(matchingKeys === keys.length) {
        this.show();
        event.preventDefault();
      }
    });

    this.$el.style.visibility = 'visible';
  },
});


code/SwitchFlitController.php
<?php

namespace SwitchFlit;

class SwitchFlitController extends \Controller
{
    private static $url_handlers = [
        '$DataObject/records' => 'getRecordsForDataObject'
    ];

    private static $allowed_actions = [
        'getRecordsForDataObject'
    ];

    private static $key_combo = 'ctrl + 75';

    private static $key_combo_mac = 'meta + 75';

    /**
     * Provide YAML-specified configuration to the browser
     * @return string JSON-encoded configuration
     */
    public function getConfig()
    {
        return (object) [
            'key_combo' => explode('+', $this->config()->key_combo),
            'key_combo_mac' => explode('+', $this->config()->key_combo_mac),
        ];
    }

    /**
     * Pulls all items from a SwitchFlitable DataObject and returns them as JSON.
     *
     * @param \SS_HTTPRequest $request The current request.
     * @return string The data in JSON format.
     *
     * @todo Clean up response handling.
     * @todo Allow custom columns? Pagination considerations?
     */
    public function getRecordsForDataObject(\SS_HTTPRequest $request)
    {
        $dataobject = $request->param('DataObject');

        if (! class_exists($dataobject)) {
            return $this->sendError('The class ' . $dataobject . ' does not exist.');
        }

        if (! in_array('DataObject', class_parents($dataobject))) {
            return $this->sendError('The class ' . $dataobject . ' is not a DataObject.');
        }

        if (! in_array('SwitchFlit\SwitchFlitable', class_implements($dataobject))) {
            return $this->sendError('The class ' . $dataobject . ' is not SwitchFlitable.');
        }

        $records = $dataobject::get();

        if (in_array('SwitchFlit\WithCustomQuery', class_implements($dataobject))) {
            $records = $dataobject::SwitchFlitQuery($records);
        }

        $data = [];

        $results = $this->prepareRecords($records);

        $response = $this->getResponse();

        $response->setStatusCode(200);
        $response->addHeader('Content-Type', 'application/json');
        $response->setBody(json_encode([
            'items' => $results,
            'config' => $this->getConfig(),
        ]));

        return $response;
    }

    public function prepareRecords($records)
    {
        $data = [];

        foreach ($records as $record) {
            if (! $record->canView()) continue;

            $data[] = [
                'title' => $record->SwitchFlitTitle(),
                'link' => $record->SwitchFlitLink()
            ];
        }

        return $data;
    }

    private function sendError($error)
    {
        $response = $this->getResponse();

        $response->setStatusCode(400);
        $response->addHeader('Content-Type', 'application/json');
        $response->setBody(json_encode([
            'errors' => [$error]
        ]));

        return $response;
    }
}
